feat(objects): Add core objects for RoomPolygon, Story, and Building

chriswmackey opened the pull request:
This commit adds a basic skeleton for the dragonfly-core objects and sets up the relationship between RoomPolygons, Stories, and Buildings.  There is still a lot left to implement until its usable (including to/from dict methods and fleshing out a number of properties that currently yield NotImplementedErrors).

@mostaphaRoudsari ,
This PR is at a state that is ready for review. I still have to add a bunch of tests but it would be good to get your input now.

mostaphaRoudsari commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+        Args:
+            name: Room name. Must be < 100 characters.
+            geometry: A single Face3D object representing the floor plate of the Room.
+            floor_to_ceiling_height: A number for the height above the floor where the
```
I think it will be simpler we just use `height` as the argument name. You can add the details to docstring.

chriswmackey replied:
That is going to create confusion with the `Story.floor_to_floor height`, then. These two types of heights are purposefully separate in the case that someone does not want to include the ceiling plenum space within the zone volume. For example, this is definitely what you want to do with a hospital, since the ventilation rate is based on the volume between floor and ceiling but there are often huge ceiling plenum spaces to accommodate all of the ventilation equipment. Other use-cases might be a 2-story building with a certain section that has a double-height gymnasium or auditorium (like a school). Having separate parameters for zone-level floor-to-ceiling height and Story floor-to-floor height will enable you to model a building like this with the Dragonfly object and not have to resort to Honeybee objects.

mostaphaRoudsari commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+                model. If None, all boundary conditions will be Outdoors or Ground
+                depending on whether ceiling of the room is below 0 (the assumed
+                ground plane).
+            glazing_ratios: A list of numbers between 0 and 1 that match the number of
```
Now that I see this I get the answer to the other comment that I left on glazing parameters class. Is there a reason not to include glazing ratio as part of the glazing parameters object?

chriswmackey replied:
There are many reasons why it's preferable to separate the amount of glass from the instructions for how to shape it. The first is ease in calculating metrics like total glazing area over a building. The second is not having to store information about how to shape the glass if you do not care about it (which is probably the case for most energy models). I was just taking a look at how floorspace.js does it, though, and I realize that they supported a few different ways to define window geometries (closer to what you have been thinking here).  However, their methods are definitely prone to errors if the user isn't checking what they put in. Maybe we should draw up a list of pros/cons and discuss.

chriswmackey replied:
I am going to resolve this sine I plan to implement it as you suggest here.

mostaphaRoudsari commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+                window geometries will be generated from the glazing_ratios. If None,
+                all window geometries will be generated by scaling the parent Face to
+                produce only one window per Face.
+            tolerance: The maximum difference between z values at which point vertices
```
what about a non-horizontal but still planar floor? It is valid as long as this input is empty or it is considered invalid?

chriswmackey replied:
Speaking realistically, I think it's invalid, which is why I plan to always use this input in the Grasshopper plugin. `None` just means that you have performed the check elsewhere and so you don't need to do it again here. FYI, I am using the same logic here with the tolerance as I do with the `Room` `__init__` method and testing whether the volume is closed:
https://github.com/ladybug-tools/honeybee-core/blob/master/honeybee/room.py#L40

mostaphaRoudsari replied:
This has a little bit of magic into it. I feel the tolerance input and enforce check for planarity can be separate.

I also think it will be helpful to be clear that it needs to be a _horizontal_ planar input.

chriswmackey replied:
I feel it doesn't have any more magic than the check for solidity in the honeybee `Room` volume but you would be right to point out that there's an additional method to check the closure of the Room volume if someone does not initialize the Room with a tolerance. So I'll do the same thing here and add a `check_horizontal` to the `Room2D` that parallels the `check_solid` property of the Room.

I'll also add some extra clarification in the docstring that the floor plane is horizontal.

mostaphaRoudsari commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+                for room_2 in room_polygons[i + 1:]:
+                    for j, seg_1 in enumerate(room_1.floor_segments):
+                        for k, seg_2 in enumerate(room_2.floor_segments):
+                            if not isinstance(room_2._boundary_conditions[k], Surface):
```
should you clean all the  boundary conditions before doing this? or at least give the option?

chriswmackey replied:
Two things:
1. I think you are misinterpreting what this line of code is doing.  This is here to catch the case where the adjacency has already been set by the first Room segment finding its adjacency to the second Room segment (there's no need to re-set the adjacency when the second segment finds the first segment again later in the iteration). So this code is really only here to save a bit of time by not having us do the proximity calculation twice.
2. We can add a method for `clean_boundary_conditions` but it's a little ambiguous what it gets "cleaned" to (ie. outdoors, ground). Also, I'm not sure when I would use it. Can you give me an example of when you need to reset all of the boundary conditions of a Room to be the same? Also is it something that really merits its own method when you can easily do this with a single for loop.

mostaphaRoudsari replied:
Not that I agree with your comments but by reading the rest of comments I think this is good for all we need now. Solving adjacency will most likely be one of the parts that will change several times in the near future.

mostaphaRoudsari commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+        """
+        for i, room_1 in enumerate(room_polygons):
+            try:
+                for room_2 in room_polygons[i + 1:]:
```
Shouldn't you check the height of the rooms first and if the height is not the same then it will not be a match?

chriswmackey replied:
I have been thinking about this and my answer for now is "No, not here". This is primarily because I know I want to support cases where you have adjacent Rooms with different floor-to-ceiling heights in the same Story, which is a very common situation. In other words, we definitely should not be keeping the boundary condition as outdoors or ground when this is the case because that's about the same as not supporting it.

For the first version of Dragonfly, I think I may make another version of this method that just sets all interior walls to be adiabatic and that's what we use for initial testing. Ultimately, though, I think I can get this to work even when the floors and/or ceilings are uneven since splitting an extrusion along its height is not a difficult task. But, in this case, all of the magic should be in the `to_honeybee()` method when we go from the 2D world over to the 3D one and the `solve_adjacency` method should be relatively unchanged from this version.

mostaphaRoudsari replied:
I think it is easier not to support rooms with different heights until the details that you mentioned are implemented otherwise this can introduce bugs for the exact cases that mentioned above.

chriswmackey replied:
I actually added full support for this since the last time I responded. As, I said, it was pretty easy to just split the wall surfaces along their height. I should note that, if someone has interior windows between adjacent walls, I enforce that the floor heights and floor-to-ceiling heights match. But, as long as there's no interior windows between adjacent segments, everything gets taken care of in the `to_honeybee()` method to ensure adjacent surfaces match.

mostaphaRoudsari commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+        self._properties = RoomPolygonProperties(self)  # properties for extensions
+
+    @property
+    def geometry(self):
```
This is a design decision but I _feel_ geometry should return the full geometry not the floor.

chriswmackey replied:
Ok but, in that case, what should we call this property? `floor_geometry`?

mostaphaRoudsari replied:
`floor_geometry` sounds good to me.

chriswmackey replied:
Ok. I will make it so.

mostaphaRoudsari commented on dragonfly/story.py:
```diff
@@ -0,0 +1,262 @@
+        return Polyface3D.from_faces(room_floors, self._tolerance)
+
+    @property
+    def outline_curves(self):
```
I suggest `polyline` instead of `outline_curves`.

chriswmackey replied:
It definitely needs to be plural since you always get a list of curves in the case that the story has a holes in the floor plate. However, when someone requests `polylines` it sounds like they might get a polyline for each `Room2D`. Not the outline of the entire floor, which is what this property is returning. You're right that `curves` is not the right term (I'm definitely stuck in Rhino SDK vocabulary here). What about `boundary_polylines`?

chriswmackey replied:
I think that I will change it to `outline_segments` since I think it is actually useful to not have them joined into a polyline. So this will return a list of `LineSegment3D` objects outlining the shape, which can be extruded to represent the context shading of the building. This way, I can implement this feature without needing to implement polylines in ladybug-geometry.

mostaphaRoudsari replied:
+1 for `outline_segments`

chriswmackey replied:
Great! I also changed it to be a method instead of a property particularly since I realized that a `tolerance` input was important for it to work correctly.

mostaphaRoudsari commented on dragonfly/story.py:
```diff
@@ -0,0 +1,262 @@
+    __slots__ = ('_room_polygons', '_floor_to_floor_height', '_is_ground_floor',
+                 '_is_top_floor', '_tolerance', '_parent')
+
+    def __init__(self, name, room_polygons, floor_to_floor_height=None,
```
From a design point of view how do you imagine to deal with solve adjacency between the rooms in the same story? Should you have a method for that in this class or do you expect one to solve adjacency before creating a story? Also how does add_room will affect that design.

chriswmackey replied:
So there are a number of things that still need to be addressed in `solve_adjacency` workflows before we really start expanding them to other areas like this. First and foremost, we have not yet determined how someone would set all interior adjacencies to be adiabatic, or change them to be a custom e+ construction or rad material, or set up air mixing between adjacent zones with air walls. I want to get some of these answered by 11/8 but we don't need to figure out this question here by the deadline since the initial tests aren't planned to be with Room2D>Story>Building workflows (essentially all zones are going to be entire floors in the Dragonfly workflows planned for 11/8).

However, to let you know what I'm thinking now, I think we can support both the option of solving adjacencies beforehand and solving all adjacencies on the `Story`. But I feel we should not be automating the solving of adjacencies each time because it's a time intensive process and people should have the freedom to set the boundary conditions to what they want.

mostaphaRoudsari replied:
Understood!

It might be helpful if you limit the options in Dragonfly and draw a line between what is possible and what is not. I don't know where that line will exactly be but having this in mind can help to focus on what makes Dargonfly different from Honeybee. Some of the more advanced features can only be available in Honeybee.

mostaphaRoudsari commented on dragonfly/building.py:
```diff
@@ -0,0 +1,241 @@
+    """
+    __slots__ = ('_unique_stories', '_multipliers')
+
+    def __init__(self, name, unique_stories, multipliers=None):
```
I feel you can include the `multiplier` as an attribute for `Story` class so it can be set in Stroy itself before adding it to building.

mostaphaRoudsari replied:
Or you might create a new class that subclasses from Story. Let's say `MultipliedStroy(Stroy)`?

chriswmackey replied:
Can you explain the advantage of that over the current implementation?

mostaphaRoudsari replied:
I think it is a cleaner implementation. Concept of multiplier can be directly applied to a story. Generally speaking whenever I see inputs like this that needs matching between two inputs I think of merging them into the reference object itself. Things can go wrong with data-matching and avoiding them makes the code less prone to errors.

chriswmackey replied:
Ok. I see what you mean. And I guess we don't have many cases where we want to have the Story divorced from its multiplier. Let me implement multipliers correctly on honeybee-core first (I agree that we definitely want them to be assigned to the `Room` there). Then, I'll revisit multipliers here and that will give me some time to mull over how everything could work here.

mostaphaRoudsari commented on dragonfly/building.py:
```diff
@@ -0,0 +1,241 @@
+        # TODO: implement footprint_geometry once we have 2D booleans in ladybug_geometry
+
+    @property
+    def context_geometry(self):
```
I know why are you adding this here to automate simulation for each building but I feel `context_geometry` shouldn't be part of the `building` class.

chriswmackey replied:
What should it be a part of then?  Or should it be a method like `get_context_representation()` instead of a property like it is here?

mostaphaRoudsari replied:
I think `ContextGeometry` can be its own object which is group of geometries which will be added to a simulation as shading objects. A `building` can be located in different `contexts`. I don't think a context necessarily belongs to a building or the other way around. They are two different independent objects.

chriswmackey replied:
So I think this is a really good suggestion if we end up adding a dragonfly `Model` object that parallels the honeybee `Model` object.  This way, people can add all of the `Buildings` they want to a Model and then add `ContextGeometry` objects to make a cohesive simulate-able unit.

However, you will see that this PR has no such `Model` object and I think that, at least for most local simulation workflows, we don't need a `Model` object in dragonfly.  Instead, we can do everything with the honeybee `Model` object. I am imagining that the user creates their dragonfly buildings and then connects them all up to a `to_honeybee` component. This component will accept settings for how to split up the input buildings into honeybee `Models`, whether this be one giant model, one `Building` per model with context representations of the other buildings, or maybe some finer-grained divisions down the line.

Whatever the final workflow is and regardless of whether we add `ContextGeometry` and a dragonfly `Model` object, we will need some way to easily get a simplified geometric representation of a `Building` so that it can inform shading of the other buildings around it. I would agree that it should not be a property of the `Building` and I would also agree that I picked a bad name here. For now, I changed it to be a method that accepts a `tolerance` input and I named the method `shade_representation`. I think this is a suitable solution for now and, if we find down the road that we really need a `ContextGeometry` object, we can always change it.

mostaphaRoudsari replied:
Thanks! You should probably open an issue for this one and link to this conversation.

chriswmackey commented on dragonfly/_base.py:
```diff
@@ -0,0 +1,63 @@
+from honeybee.typing import valid_string
+
+
+class _Base(object):
```
Ok. In that case, do you think we should do the same in honeybee-core with [_Base](https://github.com/ladybug-tools/honeybee-core/blob/master/honeybee/_base.py#L6) and [_BaseWithShade](https://github.com/ladybug-tools/honeybee-core/blob/master/honeybee/_basewithshade.py#L7)?

mostaphaRoudsari replied:
Do we really need to do that here? It depends on how much detail Dragonfly objects can take.

chriswmackey replied:
I don't understand what you mean. Weren't you the one saying that we should be changing the object names to `_BaseGeometry` and `_BaseGeometryWithShade`? Now you are asking if we really need to change them? 
Personally, I am fine with either `_Base` or `_BaseGeometry` as long as it's consistent between Dragonfly and Honeybee. Just let my know what you are suggesting.

mostaphaRoudsari replied:
My first comment was about changing the name of `_Base` to `_BaseGeometry` which I think is more appropriate.

The second comment was about  `_BaseGeometryWithShade`. Are we going to handle this much complexity in Dragonfly?

chriswmackey replied:
No we aren't going to have a need for a `_BaseWithShade` in Dragonfly and I guess I originally had an opinion that I didn't want to make dragonfly-core inconsistent with honeybee-core conventions but now I don't have a strong opinion on the matter. I'll jut make it `_BaseGeometry` since it seems you have a stronger opinion on it than I.

chriswmackey commented on requirements.txt:
```diff
@@ -1 +1 @@
-honeybee-core>=1.7.2
\ No newline at end of file
+honeybee-core
```
Done!

chriswmackey commented on dragonfly/properties.py:
```diff
@@ -0,0 +1,183 @@
+            except Exception as e:
+                import traceback
+                traceback.print_exc()
+                raise Exception('Failed to duplicate {}: {}'.format(var, e))
```
Do exceptions not automatically get logged?  If so, should we be logging every exception that we raise?

Also, we have the same exact situation in honeybee-core so whatever solution we go with should probably apply there too.

mostaphaRoudsari replied:
I didn't know the answer to this but I did a couple of tests and as far as I can see Python doesn't log an exception by default. The easiest way which is also more explicit is to use `logging.exception` which also includes the traceback by default.

https://stackoverflow.com/a/5191885/4394669

chriswmackey replied:
If exceptions aren't logged by default, then I think we need some other whole effort to go through all of the places where exceptions typically occur and log them. Right now, I don't necessarily see why this one is log-worthy and other exceptions are not but maybe we can lay out some criteria for log-worthy exceptions and implement them.

mostaphaRoudsari replied:
The difference is that now we have a logger set up! It was not the case before.

You are right that we have to fix that all over the place.

chriswmackey commented on dragonfly/properties.py:
```diff
@@ -0,0 +1,183 @@
+            except Exception as e:
+                import traceback
+                traceback.print_exc()
+                raise Exception('Failed to convert {} to a dict: {}'.format(var, e))
```
My same question applies here and whether we need to put logging on all exceptions.

chriswmackey commented on dragonfly/properties.py:
```diff
@@ -0,0 +1,183 @@
+                raise Exception('Failed to convert {} to a dict: {}'.format(var, e))
+        return base
+
+    def _load_extension_attr_from_dict(self, property_dict):
```
Ok. I have added a bit more of a description and added an `Args` tag to the docstrings that were lacking it.

chriswmackey replied:
I'll do the same for honeybee-core.

chriswmackey commented on dragonfly/glazingparameter.py:
```diff
@@ -0,0 +1,109 @@
+from honeybee.typing import float_positive
+
+
+class GlazingParameter(object):
```
You will see on the `ZonePolygon` class that the `glazing_ratios` are a separate attribute from the `glazing_parameters`.  The former just tells you how much glazing there is and the latter tells you how the specific geometry of the window surfaces should be generated.  When there are no `GlazingParameters` assigned to a given wall of a `ZonePolygon` but there is a non-zero `glazing_ratio, there will just be a single glazing surface generated that matches the ratio.  I think this is essentially the "simple glazing parameter that is based on WWR" that you are referring to, correct?

chriswmackey replied:
So I have been thinking it over and I think you are right that we should wrap the ratio into the `GlazingParameters` and plan to offer a range of different parameter objects to describe glazing geometry in different ways. In the long term, I think the same strategy will also work really well for different types of shades (overhang, fin, louvers, extruded border, etc.).

chriswmackey commented on dragonfly/glazingparameter.py:
```diff
@@ -0,0 +1,109 @@
+                the rectangle to start the apertures. Note that, if the
+                ratio is too large for the height, the ratio will take precedence
+                and the sill_height will be smaller than this value.
+            horizontal_separation: A number for the target separation between
```
The problem with allowing people to input separation between window edges and not center-lines is that it doesn't easily allow us to give precedence to glazing ratio over it. If I use center-line separation, I can easily maintain such a separation as the glazing ratio increases.  Right now, as the glazing ratio increases, I increase the width of each window until the windows touch and then I start increasing the window height beyond the specified `window_height` in order to match the `glazing_ratio`. This is at least how we decided how to implement the "glazing based on ratio component" in legacy and I think the priority of certain glazing geometry variables over others is part of what makes it so useful. I have plans to add some [alternative methods for generating windows based on a rectangular glazing Face](https://github.com/ladybug-tools/ladybug-geometry/issues/54), which might be able to accommodate requests like yours here.  But, as the link there suggests, dragonfly-core is not the right place to record a feature request like this.  Instead, an issue should be opened in ladybug-geometry and, after we address it there, we can work it into honeybee and dragonfly.

chriswmackey commented on dragonfly/glazingparameter.py:
```diff
@@ -0,0 +1,109 @@
+        horizontal_separation
+        vertical_separation
+    """
+    __slots__ = ('_room_polygons',)
```
Right you are!  Good catch.

chriswmackey commented on dragonfly/roompolygon.py:
```diff
@@ -0,0 +1,421 @@
+
+        Args:
+            name: Room name. Must be < 100 characters.
+            geometry: A single Face3D object representing the floor plate of the Room.
```
Ok. I can add `classmethods `for `from_vertices` and `from_polygon2d`. These cannot be the default `__init__` method, though, since you can't tell a hole from a boundary in a single list of vertices and I know we want to be able to support things like courtyards in Dragonfly. Also, [Polygons](https://en.wikipedia.org/wiki/Simple_polygon) by definition do not have holes in them and it is for this reason that the Polygon2D class in ladybug-geometry does not support holes (so the `from_polygon` method suffers from the same limitation as the `from_vertices` method).
Maybe down the line we will add a `PSLG` class (short for [Planar Straight-Line Graph](https://en.wikipedia.org/wiki/Planar_straight-line_graph)) to ladybug-geometry, which is the technical term for a 2D shape that can support holes. For now, though, it seems the simplest and most elegant solution in Dragonfly is to use horizontal `Face3D` objects to represent the room polygon.

chriswmackey replied:
I added the classmethods.

chriswmackey commented:
@mostaphaRoudsari ,

Sorry that I responded to your comments slowly but, at this point, I have addressed all that I could (some with structural changes to the code) and I have given written responses to the ones that need some clarification.

To summarize the major changes that happened since your last review
- I decided that your original interpretation of how `GlazingParameters` will work was better than what I had originally. These parameters should contain all of the necessary info to generate windows and not be separate from properties like glazing ratio.  this way, we can always extend the module in the future to support different ways of generating glazing from a base Face if people are not satisfied with the current methods. I added a non-glazing ratio type of `GlazingParameters` to that module to illustrate this.
- I realized that RoomPolygon was not the right name for the object since it's not really defined by a Polygon but rather by a Face that exists in 3D space and, unlike Polygons, can have holes in it. However `RoomFace` really doesn't make sense so I think going back to `Room2D` really is the best name.

chriswmackey commented:
I think that I addressed all of your comments @mostaphaRoudsari . The only two that have been left unresolved are one that you didn't respond to and one that I think should be addressed in a later PR if it becomes relevant. I just have to finish writing the tests (I'm only ~1/2 way though now) and then, with your approval, I'll merge it in.

chriswmackey commented:
All tests are done!  This one is ready to merge with your approval @mostaphaRoudsari .

chriswmackey commented:
Thanks @mostaphaRoudsari . I moved both of the open comments to issues and I am going to merge this one in for now.
